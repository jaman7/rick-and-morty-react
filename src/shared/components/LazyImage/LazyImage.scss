.img-fluid {
  max-width: 100%;
  height: auto;
  transition:
    opacity 0.3s ease-in-out,
    transform 0.3s ease-in-out;
  will-change: opacity, transform;
  overflow: hidden;
  position: relative;

  &::before {
    content: "";
    position: absolute;
    inset: 0;
    background: linear-gradient(90deg, #e0e0e0 0%, #f0f0f0 50%, #e0e0e0 100%);
    background-size: 200% 100%;
    animation: shimmer 1.5s infinite;
    z-index: 1;
  }

  &.lazyloaded::before {
    display: none;
  }
}

.lazyload,
.lazyloading {
  opacity: 0;
  transform: scale(0.9);
  filter: blur(2px);
  background-color: #e0e0e0;
}

.lazyloaded {
  opacity: 1;
  transform: scale(1);
  filter: none;
}

.lazy-placeholder {
  width: 100%;
  padding-top: 66%; // zale≈ºnie od proporcji
  background: linear-gradient(90deg, #ccc 25%, #ddd 37%, #ccc 63%);
  background-size: 400% 100%;
  animation: shimmer 1.2s ease-in-out infinite;
  border-radius: 8px;
}

@keyframes shimmer {
  0% {
    background-position: -200% 0;
  }

  100% {
    background-position: 200% 0;
  }
}
