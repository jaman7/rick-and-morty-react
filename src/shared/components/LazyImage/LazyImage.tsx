import { useState, useEffect, useRef } from 'react';
import './LazyImage.scss';

export interface ILazyImage {
  id?: string;
  className?: string;
  src?: string | null;
  alt?: string;
  onClick?: () => void;
  placeholderSrc?: string;
}

const DEFAULT_LOGO = 'https://upload.wikimedia.org/wikipedia/commons/a/ac/No_image_available.svg';

const LazyImage: React.FC<ILazyImage> = ({ id, className, src = null, alt, onClick, placeholderSrc = '' }) => {
  const [loaded, setLoaded] = useState(false);
  const [error, setError] = useState(false);
  const imgRef = useRef<HTMLImageElement>(null);

  useEffect(() => {
    const observer = new IntersectionObserver(
      ([entry]) => {
        if (entry.isIntersecting) {
          observer.disconnect();
        }
      },
      { threshold: 0.1 }
    );

    if (imgRef.current) observer.observe(imgRef.current);
    return () => observer.disconnect();
  }, []);

  return (
    <>
      {!loaded && !error && (
        <img
          src={placeholderSrc || 'data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=='}
          alt=""
          aria-hidden="true"
          role="presentation"
        />
      )}
      {!loaded && <div className="lazy-placeholder" data-testid="lazy-placeholder" aria-hidden="true" />}
      <img
        id={id}
        loading="lazy"
        src={error || !src ? DEFAULT_LOGO : src}
        alt={alt}
        ref={imgRef}
        onLoad={() => setLoaded(true)}
        onError={() => setError(true)}
        className={`img-fluid ${className} ${loaded ? 'lazyloaded' : 'lazyloading'}`}
        onClick={onClick}
      />
    </>
  );
};

export default LazyImage;
