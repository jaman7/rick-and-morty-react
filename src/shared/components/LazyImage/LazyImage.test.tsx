import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import LazyImage, { ILazyImage } from './LazyImage';
import { describe, it, expect, vi } from 'vitest';

describe('LazyImage', () => {
  const defaultProps: ILazyImage = {
    src: 'https://example.com/image.jpg',
    alt: 'Test image',
    className: 'custom-class',
  };

  beforeAll(() => {
    Object.defineProperty(global.Image.prototype, 'src', {
      set(src) {
        this._src = src;
        setTimeout(() => {
          this.onload?.();
        }, 100);
      },
      get() {
        return this._src;
      },
    });
  });

  it('renders placeholder before image loads', async () => {
    render(<LazyImage {...defaultProps} />);

    const placeholder = screen.queryByTestId('lazy-placeholder');
    expect(placeholder).toBeInTheDocument();

    const image = screen.getByAltText('Test image');
    expect(image).toHaveClass('lazyloading');

    fireEvent.load(image);

    await waitFor(() => {
      expect(screen.queryByTestId('lazy-placeholder')).not.toBeInTheDocument();
    });

    expect(image).toHaveClass('lazyloaded');
  });

  it('loads image and applies lazyloaded class on load', () => {
    render(<LazyImage {...defaultProps} />);
    const img = screen.getAllByRole('img')[0];
    fireEvent.load(img);
    expect(img).toHaveClass('lazyloaded');
  });

  it('shows fallback image on error', () => {
    render(<LazyImage {...defaultProps} />);
    const img = screen.getAllByRole('img')[0];
    fireEvent.error(img);
    expect(img).toHaveAttribute('src', 'https://upload.wikimedia.org/wikipedia/commons/a/ac/No_image_available.svg');
  });

  it('calls onClick when clicked', () => {
    const onClick = vi.fn();
    render(<LazyImage {...defaultProps} onClick={onClick} />);
    const img = screen.getAllByRole('img')[0];
    fireEvent.click(img);
    expect(onClick).toHaveBeenCalled();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<LazyImage {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
