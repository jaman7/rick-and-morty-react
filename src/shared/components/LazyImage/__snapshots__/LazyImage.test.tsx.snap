// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LazyImage > matches snapshot 1`] = `
<DocumentFragment>
  <img
    alt=""
    aria-hidden="true"
    role="presentation"
    src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
  />
  <div
    aria-hidden="true"
    class="lazy-placeholder"
    data-testid="lazy-placeholder"
  />
  <img
    alt="Test image"
    class="img-fluid custom-class lazyloading"
    loading="lazy"
    src="https://example.com/image.jpg"
  />
</DocumentFragment>
`;
